### PR TITLE
MULE-7081: SFTP sizeCheckWaitTime should be applied per poll cycle in…

### PR DESCRIPTION
**Pull request by facundovs**
…stead of per file.

### CODE REVIEW

File: transports/sftp/src/main/java/org/mule/transport/sftp/SftpReceiverRequesterUtil.java [hunk 1]
```diff
@@ -143,6 +155,24 @@ public SftpReceiverRequesterUtil(ImmutableEndpoint endpoint)
         }
     }
 
+    /**
+     * Allows to check if an SFTP Connection can be established.
+     * @throws Exception if the connection can not be established.
+     */
+    public void checkSFTPConnection () throws Exception
```
Comment (pablokraan) [inline transports/sftp/src/main/java/org/mule/transport/sftp/SftpReceiverRequesterUtil.java]: Can be package protected

File: transports/sftp/src/main/java/org/mule/transport/sftp/SftpReceiverRequesterUtil.java [hunk 2]
```diff
@@ -83,14 +85,24 @@ public SftpReceiverRequesterUtil(ImmutableEndpoint endpoint)
...
             // Get size check parameter
             long sizeCheckDelayMs = sftpUtil.getSizeCheckWaitTime();
 
-            String[] files = client.listFiles();
+            // The files that has the same size after the sizeCheckDelay.
+            List<String> stableFiles = null;
+
+            if (sizeCheckDelayMs > 0)
+            {
+                stableFiles = getStableFiles(files, client, sizeCheckDelayMs);
```
Comment (pablokraan) [inline transports/sftp/src/main/java/org/mule/transport/sftp/SftpReceiverRequesterUtil.java]: Can you filter the original files list instead of using a different list?

File: transports/sftp/src/main/java/org/mule/transport/sftp/SftpReceiverRequesterUtil.java [hunk 3]
```diff
@@ -143,6 +155,24 @@ public SftpReceiverRequesterUtil(ImmutableEndpoint endpoint)
         }
     }
 
+    /**
+     * Allows to check if an SFTP Connection can be established.
+     * @throws Exception if the connection can not be established.
```
Comment (pablokraan) [inline transports/sftp/src/main/java/org/mule/transport/sftp/SftpReceiverRequesterUtil.java]: Add extra line between method and throws descriptions

File: transports/sftp/src/main/java/org/mule/transport/sftp/SftpReceiverRequesterUtil.java [hunk 4]
```diff
@@ -333,70 +363,94 @@ private InputStream archiveFile(SftpInputStream is, File archiveFile) throws IOE
      * @param client instance of StftClient
      * @param fileAge How old the file should be to be considered "old" and not
      *            changed
-     * @param sizeCheckDelayMs Wait time (in ms) between size-checks to determine if
-     *            a file is ready to be processed.
      * @return true if the file has changed
      * @throws Exception Error
```
Comment (pablokraan) [inline transports/sftp/src/main/java/org/mule/transport/sftp/SftpReceiverRequesterUtil.java]: Update javadoc accordingly to your changes

File: transports/sftp/src/main/java/org/mule/transport/sftp/SftpReceiverRequesterUtil.java [hunk 5]
```diff
@@ -333,70 +363,94 @@ private InputStream archiveFile(SftpInputStream is, File archiveFile) throws IOE
...
+        List <String> stableFiles = new ArrayList<>(fileNames.length);
+        Map<String, Long> sizes = new HashMap<>();
+
+        if (logger.isDebugEnabled())
         {
-            if (logger.isDebugEnabled())
+            logger.debug("Perform size check with a delay of: " + sizeCheckDelayMs + " ms.");
+        }
+
+        for (String fileName : fileNames)
```
Comment (pablokraan) [inline transports/sftp/src/main/java/org/mule/transport/sftp/SftpReceiverRequesterUtil.java]: Extract this to a method "getFileTimeStamps" or similar, then refactor the code to do something like:

a = getFileTimeStamps
sleep
b = getFileTimeStamps
filter a after comparing against b

File: transports/sftp/src/main/java/org/mule/transport/sftp/SftpReceiverRequesterUtil.java [hunk 6]
```diff
@@ -434,4 +488,10 @@ private boolean isOldFile(String fileName, SftpClient client, long fileAge) thro
             return false;
         }
     }
+
+    protected SftpUtil createSftpUtil (ImmutableEndpoint endpoint)
```
Comment (pablokraan) [inline transports/sftp/src/main/java/org/mule/transport/sftp/SftpReceiverRequesterUtil.java]: Add javadoc

File: transports/sftp/src/test/java/org/mule/transport/sftp/SftpReceiverRequesterUtilTestCase.java [hunk 7]
```diff
@@ -58,4 +75,138 @@ public void clientIsReleasedWhenRetrievingFails() throws Exception
...
+     * Verifies an stable file (i.e. a file whose size doesn't change) is returned when fileAge is disabled.
+     */
+    @Test
+    public void testReturnsStableFile() throws Exception
+    {
+        final String stableFile = "file1";
+        when(connector.getFileAge()).thenReturn(-1L);
+        when(sftpClient.listFiles()).thenReturn(new String[] {stableFile});
+        when(sftpClient.getSize(stableFile)).thenReturn(1024L);
+        SftpReceiverRequesterUtil requesterUtil = new TestSftpReceiverRequesterUtil(endpoint);
```
Comment (pablokraan) [inline transports/sftp/src/test/java/org/mule/transport/sftp/SftpReceiverRequesterUtilTestCase.java]: Add extra lines to separate the different parts of the tests to make them more readeable (setup, execute, assert)

File: transports/sftp/src/test/java/org/mule/transport/sftp/SftpReceiverRequesterUtilTestCase.java [hunk 8]
```diff
@@ -58,4 +75,138 @@ public void clientIsReleasedWhenRetrievingFails() throws Exception
             verify(connector).releaseClient(endpoint, sftpClient);
         }
     }
+
+    /**
+     * Verifies an stable file (i.e. a file whose size doesn't change) is returned when fileAge is disabled.
+     */
+    @Test
+    public void testReturnsStableFile() throws Exception
```
Comment (pablokraan) [inline transports/sftp/src/test/java/org/mule/transport/sftp/SftpReceiverRequesterUtilTestCase.java]: Make the test method names more clear instead of adding a javadoc

File: transports/sftp/src/main/java/org/mule/transport/sftp/SftpReceiverRequesterUtil.java [hunk 9]
```diff
@@ -325,78 +337,89 @@ private InputStream archiveFile(SftpInputStream is, File archiveFile) throws IOE
...
      *            changed
-     * @param sizeCheckDelayMs Wait time (in ms) between size-checks to determine if
-     *            a file is ready to be processed.
-     * @return true if the file has changed
+     * @return true if the file meets the conditions to be processed.
      * @throws Exception Error
      */
-    protected boolean canProcessFile(String fileName, SftpClient client, long fileAge, long sizeCheckDelayMs)
-        throws Exception
+    protected boolean canProcessFile(String fileName, SftpClient client, boolean checkFileAge, long fileAge)
```
Comment (pablokraan) [inline transports/sftp/src/main/java/org/mule/transport/sftp/SftpReceiverRequesterUtil.java]: It makes no sense to pass a checkFileAge=false and a fileAge, so only call this method when it is required and remove the checkFileAge parameter

File: transports/sftp/src/main/java/org/mule/transport/sftp/SftpReceiverRequesterUtil.java [hunk 10]
```diff
@@ -325,78 +337,89 @@ private InputStream archiveFile(SftpInputStream is, File archiveFile) throws IOE
...
+            throws Exception
     {
-        if (fileAge > 0 && !isOldFile(fileName, client, fileAge))
-        {
-            return false;
-        }
 
-        if (sizeCheckDelayMs > 0 && isSizeModified(fileName, client, sizeCheckDelayMs))
+        if (checkFileAge && fileAge > 0)
```
Comment (pablokraan) [inline transports/sftp/src/main/java/org/mule/transport/sftp/SftpReceiverRequesterUtil.java]: This condition seems redundant. If checkFileAge=true, then fileAge must be greater than zero

File: transports/sftp/src/main/java/org/mule/transport/sftp/SftpReceiverRequesterUtil.java [hunk 11]
```diff
@@ -325,78 +337,89 @@ private InputStream archiveFile(SftpInputStream is, File archiveFile) throws IOE
...
             }
+        }
 
-            long fileSize1 = client.getSize(fileName);
-            Thread.sleep(sizeCheckDelayMs);
-            long fileSize2 = client.getSize(fileName);
+        return stableFiles;
+    }
+
+
```
Comment (pablokraan) [inline transports/sftp/src/main/java/org/mule/transport/sftp/SftpReceiverRequesterUtil.java]: Remove extra line

File: transports/sftp/src/main/java/org/mule/transport/sftp/SftpReceiverRequesterUtil.java [hunk 12]
```diff
@@ -325,78 +337,89 @@ private InputStream archiveFile(SftpInputStream is, File archiveFile) throws IOE
...
-            if (fileSize1 == fileSize2)
+        for (String fileName : fileNames)
+        {
+            try
             {
-                if (logger.isDebugEnabled())
-                {
-                    logger.debug("File is stable (not growing), ready for retrieval: " + fileName);
-                }
+                long size = client.getSize(fileName);
```
Comment (pablokraan) [inline transports/sftp/src/main/java/org/mule/transport/sftp/SftpReceiverRequesterUtil.java]: Inline this variable

File: transports/sftp/src/main/java/org/mule/transport/sftp/SftpReceiverRequesterUtil.java [hunk 13]
```diff
@@ -325,78 +337,89 @@ private InputStream archiveFile(SftpInputStream is, File archiveFile) throws IOE
...
+                sizes.put(fileName, size);
             }
-            else
+            catch (IOException e)
             {
                 if (logger.isDebugEnabled())
                 {
-                    logger.debug("File is growing, deferring retrieval: " + fileName);
+                    logger.debug(String.format("Cannot check if size of file '%s' was modified or not", fileName));
```
Comment (pablokraan) [inline transports/sftp/src/main/java/org/mule/transport/sftp/SftpReceiverRequesterUtil.java]: Static import

File: transports/sftp/src/main/java/org/mule/transport/sftp/SftpConnector.java [hunk 14]
```diff
@@ -496,7 +496,10 @@ public long getFileAge()
     public void setFileAge(long fileAge)
     {
         this.fileAge = fileAge;
-        this.checkFileAge = true;
+        if(fileAge > 0)
```
Comment (pablokraan) [inline transports/sftp/src/main/java/org/mule/transport/sftp/SftpConnector.java]: Fix code style (check all your changes)

File: transports/sftp/src/main/java/org/mule/transport/sftp/SftpReceiverRequesterUtil.java [hunk 15]
```diff
@@ -143,6 +149,25 @@ public SftpReceiverRequesterUtil(ImmutableEndpoint endpoint)
         }
     }
 
+    /**
+     * Allows to check if an SFTP Connection can be established.
```
Comment (pablokraan) [inline transports/sftp/src/main/java/org/mule/transport/sftp/SftpReceiverRequesterUtil.java]: Allows to check > Checks

File: transports/sftp/src/main/java/org/mule/transport/sftp/SftpReceiverRequesterUtil.java [hunk 16]
```diff
@@ -324,79 +349,66 @@ private InputStream archiveFile(SftpInputStream is, File archiveFile) throws IOE
...
-     *            a file is ready to be processed.
-     * @return true if the file has changed
-     * @throws Exception Error
+     * Filters the given files evaluating if their size changed after <code>sizeCheckDelayMs</code>.
+     *
+     * @param fileNames the name of the files to evaluate
+     * @param client an SftpClient
+     * @param sizeCheckDelayMs the delay time in milliseconds.
+     * @return a list with the files whose size didn't change after the <code>sizeCheckDelayMs<code/>.
+     * @throws InterruptedException
```
Comment (pablokraan) [inline transports/sftp/src/main/java/org/mule/transport/sftp/SftpReceiverRequesterUtil.java]: Add a comment explaining why throws this exception

File: transports/sftp/src/main/java/org/mule/transport/sftp/SftpReceiverRequesterUtil.java [hunk 17]
```diff
@@ -324,79 +349,66 @@ private InputStream archiveFile(SftpInputStream is, File archiveFile) throws IOE
...
+    List<String> getStableFiles(List <String> fileNames, SftpClient client, long sizeCheckDelayMs) throws InterruptedException
     {
-        if (fileAge > 0 && !isOldFile(fileName, client, fileAge))
-        {
-            return false;
-        }
+        List <String> stableFiles = new ArrayList<>(fileNames.size());
 
-        if (sizeCheckDelayMs > 0 && isSizeModified(fileName, client, sizeCheckDelayMs))
+        if (logger.isDebugEnabled())
```
Comment (pablokraan) [inline transports/sftp/src/main/java/org/mule/transport/sftp/SftpReceiverRequesterUtil.java]: This logs does not seem to add much value

File: transports/sftp/src/main/java/org/mule/transport/sftp/SftpReceiverRequesterUtil.java [hunk 18]
```diff
@@ -324,79 +349,66 @@ private InputStream archiveFile(SftpInputStream is, File archiveFile) throws IOE
...
         }
 
-        return true;
-    }
 
-    private boolean isSizeModified(String fileName, SftpClient client, long sizeCheckDelayMs) throws IOException, InterruptedException
-    {
-        try
+        Map<String, Long> fileSizesBeforeDelay = getFileTimeStamps(fileNames, client);
+        Thread.sleep(sizeCheckDelayMs);
```
Comment (pablokraan) [inline transports/sftp/src/main/java/org/mule/transport/sftp/SftpReceiverRequesterUtil.java]: static import

File: transports/sftp/src/main/java/org/mule/transport/sftp/SftpReceiverRequesterUtil.java [hunk 19]
```diff
@@ -324,79 +349,66 @@ private InputStream archiveFile(SftpInputStream is, File archiveFile) throws IOE
...
-                }
+                sizes.put(fileName, client.getSize(fileName));
             }
-            else
+            catch (IOException e)
             {
                 if (logger.isDebugEnabled())
                 {
-                    logger.debug("File is growing, deferring retrieval: " + fileName);
+                    logger.debug(format("Cannot check if size of file '%s' was modified or not", fileName));
```
Comment (pablokraan) [inline transports/sftp/src/main/java/org/mule/transport/sftp/SftpReceiverRequesterUtil.java]: Error must say that cannot obtain the file size

File: transports/sftp/src/test/resources/mule-sftp-connection-returned-to-pool-on-error-config.xml [hunk 20]
```diff
@@ -15,7 +15,7 @@
     <sftp:connector name="sftpPooledConnector"
                     pollingFrequency="1000"
                     autoDelete="false"
-                    sizeCheckWaitTime="1"
+                    fileAge="1"
```
Comment (pablokraan) [inline transports/sftp/src/test/resources/mule-sftp-connection-returned-to-pool-on-error-config.xml]: Why was this change required?

Comment (facundovs) [inline transports/sftp/src/test/resources/mule-sftp-connection-returned-to-pool-on-error-config.xml]: The related test case has a latch that is counted down when the file is processed (before this test case overrode the canProcessFile method in a Test MessageReceiver). Due to these changes, I have to replace the canProcessFile method (that doesn't exist anymore) with the isOldFile method, and the sizeCheckWaitTime with the fileAge attribute.

File: transports/sftp/src/main/java/org/mule/transport/sftp/SftpReceiverRequesterUtil.java [hunk 21]
```diff
@@ -143,6 +150,25 @@ public SftpReceiverRequesterUtil(ImmutableEndpoint endpoint)
         }
     }
 
+    /**
+     * Checks if an SFTP Connection can be established.
+     *
+     * @throws Exception if the connection can not be established.
+     */
+    void checkSFTPConnection () throws Exception
```
Comment (pablokraan) [inline transports/sftp/src/main/java/org/mule/transport/sftp/SftpReceiverRequesterUtil.java]: Fix code style
